src/libLookup/Lookup.cpp

```diff
@@ -204,26 +204,35 @@ void Lookup::GetInitialBlocksAndShardingStructure() {
     }
     LOG_GENERAL(INFO,
                 "TxBlockNum " << txBlockNum << " DSBlockNum: " << dsBlockNum);
+    {
+      std::unique_lock<std::mutex> cv_lk(m_mutexCvSetRejoinRecovery);
+      m_rejoinRecoverySignal = false;
+    }
     ComposeAndSendGetDirectoryBlocksFromSeed(
         m_mediator.m_blocklinkchain.GetLatestIndex() + 1, true,
         LOOKUP_NODE_MODE);
     GetTxBlockFromSeedNodes(txBlockNum, 0);
     std::unique_lock<std::mutex> cv_lk(m_mutexCvSetRejoinRecovery);
-    if (m_mediator.m_lookup->cv_setRejoinRecovery.wait_for(
-            cv_lk, std::chrono::seconds(NEW_NODE_SYNC_INTERVAL)) !=
-        std::cv_status::timeout) {
+    if (!cv_setRejoinRecovery.wait_for(
+            cv_lk, std::chrono::seconds(NEW_NODE_SYNC_INTERVAL), [this]() {
+              return m_rejoinRecoverySignal.load();
+            })) {
       if (m_rejoinInProgress) {
         break;
       }
     }
   }
   if (!m_rejoinInProgress) {
     // Ask for the sharding structure from lookup
+    {
+      std::unique_lock<std::mutex> cv_lk(m_mutexShardStruct);
+      m_shardStructSignal = false;
```

Comment:
Same as above: why not just set after line 223/232 below?

Reply:
Same as above: we need to keep the mutex unlocked when the network communication request is initiated, because the response may come sooner than we expect (when this thread is interrupted for example, and only scheduled later).